In [87]:
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы plant<b>

In [88]:
plant = wordnet.synsets('plant')
for ss in plant:
    print(ss, ss.definition())

Synset('plant.n.01') buildings for carrying on industrial labor
Synset('plant.n.02') (botany) a living organism lacking the power of locomotion
Synset('plant.n.03') an actor situated in the audience whose acting is rehearsed but seems spontaneous to the audience
Synset('plant.n.04') something planted secretly for discovery by another
Synset('plant.v.01') put or set (seeds, seedlings, or plants) into the ground
Synset('implant.v.01') fix or set securely or deeply
Synset('establish.v.02') set up or lay the groundwork for
Synset('plant.v.04') place into a river
Synset('plant.v.05') place something or someone in a certain position in order to secretly observe or deceive
Synset('plant.v.06') put firmly in the mind


<b>2) Найти определение для лексемы plant в значении (а) "завод" и в значении (b) "растение"</b>

In [89]:
print(plant[0], plant[0].definition())
print(plant[1], plant[1].definition())

Synset('plant.n.01') buildings for carrying on industrial labor
Synset('plant.n.02') (botany) a living organism lacking the power of locomotion


<b>3) Найдите два произвольных контекста для слова plant в значениях (a) "завод" и (b) "растение"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [90]:
sent1 = 'County Park Nursery specialises in plants from New Zealand, many grown from seed collected on plant hunting expeditions.'
sent2 = 'Coun Mark Harris, council leader, challenged Labour to spell out how they would tackle the problem and reveal where they would site a waste treatment plant.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['County', 'Park', 'Nursery', 'specialises', 'in', 'plants', 'from', 'New', 'Zealand', 'many', 'grown', 'from', 'seed', 'collected', 'on', 'plant', 'hunting', 'expeditions']
['Coun', 'Mark', 'Harris', 'council', 'leader', 'challenged', 'Labour', 'to', 'spell', 'out', 'how', 'they', 'would', 'tackle', 'the', 'problem', 'and', 'reveal', 'where', 'they', 'would', 'site', 'a', 'waste', 'treatment', 'plant']


In [91]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'plant').definition())
print(lesk(sent2_tokens, 'plant').definition())

put firmly in the mind
place something or someone in a certain position in order to secretly observe or deceive


Алгоритм Леска ошибся в обоих случаях. Более того, он ошибся даже с частью речи. Попробуем ограничить часть речи существительными.

In [92]:
print(lesk(sent1_tokens, 'plant', 'n').definition())
print(lesk(sent2_tokens, 'plant', 'n').definition())

an actor situated in the audience whose acting is rehearsed but seems spontaneous to the audience
an actor situated in the audience whose acting is rehearsed but seems spontaneous to the audience


Не помогло :(

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [93]:
for ss in plant[0].hypernyms():
    print(ss, ss.definition())
for ss in plant[1].hypernyms():
    print(ss, ss.definition())

Synset('building_complex.n.01') a whole structure (as a building) made up of interconnected or related structures
Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently


<b>5) Вычислите наименьшее расстояние между значением plant "завод" и значениями лексемы industry, а также plant "растение" и значениями лексемы leaf.
Найти min (d(plant: "завод", industry), d(plant: "завод", leaf)), а также min (d(plant: "растение", industry), d(plant: "растение", leaf))</b>

In [94]:
industry = wordnet.synsets('industry')
leaf = wordnet.synsets('leaf')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(plant: "завод", industry)
dist1 = get_dist_sim(plant[0], industry)[0]
print('min d(plant: "завод", industry): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(industry[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(plant: "завод", leaf)
dist2 = get_dist_sim(plant[0], leaf)[0]
print('min d(plant: "завод", leaf): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(leaf[dist2.index(min(dist2))].definition()))

# min d(plant: "растение", industry)
dist3 = get_dist_sim(plant[1], industry)[0]
print('min d(plant: "растение", industry): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(industry[dist3.index(min(dist3))].definition()))

# min d(plant: "растение", leaf)
dist4 = get_dist_sim(plant[1], leaf)[0]
print('min d(plant: "растение", leaf): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(leaf[dist4.index(min(dist4))].definition()))

print('min (d(plant: "завод", industry), d(plant: "завод", leaf)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(plant: "растение", industry), d(plant: "растение", leaf)): {}'.format(min(min(dist3), min(dist4))))

min d(plant: "завод", industry): 13
closest lemma definition: persevering determination to perform a task

min d(plant: "завод", leaf): 8
closest lemma definition: the main organ of photosynthesis and transpiration in higher plants

min d(plant: "растение", industry): 12
closest lemma definition: persevering determination to perform a task

min d(plant: "растение", leaf): 7
closest lemma definition: the main organ of photosynthesis and transpiration in higher plants

min (d(plant: "завод", industry), d(plant: "завод", leaf)): 8
min (d(plant: "растение", industry), d(plant: "растение", leaf)): 7


Кратчайшее расстояние от завода до industry - 13 (и у ближайшей леммы industry весьма не тривиальное для данного задания значение), до leaf - 8. Получается, что до leaf расстояние короче, что кажется странным.
Кратчайшее расстояние от растения до industry - 12, до leaf - 7. До leaf расстояние короче, слава богу.

Если имелось в виду посчитать не расстояние, а сходство путей на основе кратчайшего расстояния (минимальный путь --> максимальное сходство), вот оно:

In [95]:
# max s(plant: "завод", industry)
sim1 = get_dist_sim(plant[0], industry)[1]
print('max s(plant: "завод", industry): {}'.format(max(sim1)))
print('closest lemma definition: {}\n'.format(industry[sim1.index(max(sim1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# max s(plant: "завод", leaf)
sim2 = get_dist_sim(plant[0], leaf)[1]
print('max s(plant: "завод", leaf): {}'.format(max(sim2)))
print('closest lemma definition: {}\n'.format(leaf[sim2.index(max(sim2))].definition()))

# max s(plant: "растение", industry)
sim3 = get_dist_sim(plant[1], industry)[1]
print('max s(plant: "растение", industry): {}'.format(min(sim3)))
print('closest lemma definition: {}\n'.format(industry[sim3.index(max(sim3))].definition()))

# max s(plant: "растение", leaf)
sim4 = get_dist_sim(plant[1], leaf)[1]
print('max s(plant: "растение", leaf): {}'.format(max(sim4)))
print('closest lemma definition: {}\n'.format(leaf[sim4.index(max(sim4))].definition()))

print('max (s(plant: "завод", industry), s(plant: "завод", leaf)): {}'.format(max(max(sim1), max(sim2))))
print('max (s(plant: "растение", industry), s(plant: "растение", leaf)): {}'.format(max(max(sim3), max(sim4))))

max s(plant: "завод", industry): 0.07142857142857142
closest lemma definition: persevering determination to perform a task

max s(plant: "завод", leaf): 0.1111111111111111
closest lemma definition: the main organ of photosynthesis and transpiration in higher plants

max s(plant: "растение", industry): 0.06666666666666667
closest lemma definition: persevering determination to perform a task

max s(plant: "растение", leaf): 0.125
closest lemma definition: the main organ of photosynthesis and transpiration in higher plants

max (s(plant: "завод", industry), s(plant: "завод", leaf)): 0.1111111111111111
max (s(plant: "растение", industry), s(plant: "растение", leaf)): 0.125


Результат тот же (что ожидаемо, потому что используется фактически одна и та же мера). И завод, и растение ближе к leaf.

<b>6) Вычислить двумя разными способами расстояние:
d(plant: "растение", rattlesnake's master) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

А вот здесь же точно речь идет о сходстве/близости значений, а не о стандартном расстоянии? (Потому что стандартное расстояние - кратчайший путь - трудно считать несколькими способами.)

Synset rattlesnake's master не находится, по-видимому, из-за апострофа или пробела. Заменим пробел на подчеркивание - всё ок.

In [103]:
master = wordnet.synsets("rattlesnake's_master")
for ss in master:
    print(ss, ss.definition())

Synset('rattlesnake_master.n.01') coarse prickly perennial eryngo of United States thought to cure rattlesnake bite


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [107]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(plant[1].path_similarity(master[0]))

#Leacock-Chodorow Similarity
print(plant[1].lch_similarity(master[0]))

#Wu-Palmer Similarity
print(plant[1].wup_similarity(master[0]))
    
#Resnik Similarity
print(plant[1].res_similarity(master[0], ic))
    
#Jiang-Conrath Similarity
print(plant[1].jcn_similarity(master[0], ic))
    
#Lin Similarity
print(plant[1].lin_similarity(master[0], ic))


0.3333333333333333
2.538973871058276
0.875
5.948206741643484
1e-300
1.1896413483286967e-299


In [112]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [113]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [111]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>растение/rattlesnake's master</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.3333333333</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>2.538973871</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.875</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>5.948206742</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>1.00E-300</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>1.19E-299</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>

Rattlesnake's master просто является гипонимом растения. Как его сравнивать с organism/whole не совсем понятно, потому что во втором случае нет таких прямых отношений. За исключением последних двух индексов у этой пары самые большие показатели схожести.

Интуитивно кажется, что оба значения организма (a living thing that has (or can develop) the ability to act or function independently; a system considered analogous in structure or function to a living body) должны быть больше похожи на второе значение целого (an assemblage of parts that is regarded as a single entity), чем на первое (all of something including all its component elements or parts). То есть четвертый и шестой столбцы в таблице должны быть больше, чем третий и пятый. Хотя второе значение организма и первое значение целого тоже близки.

В четвертом столбце действительно самые большие индексы среди организма/целого. По последним двум индексам эта пара превосходит и растение/rattlesnake's master. 

Показатели в шестом же столбце больше третьего и пятого только для первых двух индексов (по Wu-Palmer Similarity разница небольшая). Первые два индекса, кстати, дают более или менее пропорциональные величины, и сравнение по ним дает одинаковые результаты (это значит, что глубина наших значений в данном случае не влияет на результат).

У пары растение/rattlesnake's master по последним двум индексам значения очень близкие к нулю. Кажется, что это противоречит интуиции. Resnik Similarity и Lin Similarity в третьем и шестом столбце дают нули. Непонятно, говорит ли это о полной несхожести значений или это специфика индекса/корпуса/данных.

А вообще, значения индексов в основном довольно маленькие. Не совсем понятно, насколько целесообразно сравнение сотых и тысячных долей.

Если учитывать всё сразу, первые три индекса более или менее отражают интуитивное представление о близости слов.